homeassistant/components/mqtt/client.py

```diff
@@ -594,59 +615,38 @@ def _client_unsubscribe(topic: str) -> int:
             _raise_on_error(result)
             return mid
 
-        async with self._paho_lock:
-            if self._is_active_subscription(topic):
-                # Other subscriptions on topic remaining - don't unsubscribe.
-                return
+        if self._is_active_subscription(topic):
+            # Other subscriptions on topic remaining - don't unsubscribe.
+            return
+
+        async with self._subscriptions_lock:
+            if topic in self._pending_subscriptions:
+                del self._pending_subscriptions[topic]
```

Comment:
This doesn't seem right. I think we need to count the number of pending subscriptions?

Reply:
No, right above you count *active* subscription, that doesn't include *pending* subscriptions, does it?

Reply:
That's Here's the sequence:

1. A subscribes to foo
  - Subscription (class) for foo for A is created
  - pending for foo is added.
2. B subscribes to foo
  - Subscription for foo for B is created
3. A unsubscribes
  - Subscription for foo for A is removed.
  - pending for foo should NOT be removed (subscription for B exists).
4. B unsubscribes
  - Subscription for foo for B is removed.
  - pending for foo should be removed.

This can all happen before the pending subscription is processed due to the delay.
Without this code, the subscription entry for A and B would be removed.  But the pending subscription would still exist and we would still subscribe to that topic.

If I just blindly removed the pending subscription, and either B or A had not unsubscribed, we would never perform a subscription that was still needed.